### PR TITLE
Add ADADELTA algorithm

### PR DESCRIPTION
**Pull request by urikz**
Just one more optimization algorithm. I wonder if that will be helpful for others.

Paper: http://www.matthewzeiler.com/pubs/googleTR2012/googleTR2012.pdf

Test plan:

    $ luarocks make optim-1.0.5-0.rockspec
    $ cd test
    $ luarocks make

    $ th test_adadelta.lua
    Rosenbrock test
    x=
     0.9932
     0.9861
    [torch.DoubleTensor of dimension 2]    

    fx=
    1	1
    1001	0.85674832729526
    2001	0.65380034382089
    3001	0.45083429533603
    4001	0.27774607313817
    5001	0.14830309667068
    6001	0.06446454527491
    7001	0.019846854100354
    8001	0.0029878163396553
    9001	0.00021361750977478
    10001	6.034663542954e-05    

    $ th test_adagrad.lua
    Rosenbrock test
    x=
     1.0000
     1.0000
    [torch.DoubleTensor of dimension 2]    

    fx=
    1	1
    1001	0.0096629336429242
    2001	0.0011605963642662
    3001	0.00017238390470642
    4001	0.00037591610790092
    5001	4.4145804955209e-06
    6001	7.1646069394023e-07
    7001	1.1654181321944e-07
    8001	1.8974400763049e-08
    9001	3.090397609615e-09
    10001	5.0341388774798e-10

### CODE REVIEW

**Comment by urikz** [comment]
@soumith , mind to take a quick look?

**Comment by soumith** [comment]
:clone() is doing a malloc at every iteration. this can lead to weird memory consumption that you will find hard to debug. can you remove the need for :clone() at all places in this file?

**Comment by soumith** [comment]
otherwise, this looks good to me.

**Comment by urikz** [comment]
@soumith , I used `typeAs` + `resizeAs` + `add`, not sure if that's what you meant. But I feel that `clone` is more straightforward and will potentially costs less operations.

**Comment by soumith** [comment]
@urikz I think you missed my point completely :), typeAs + resizeAs also do malloc, the idea is to not do malloc at every iteration, but keep temporary buffers in the state itself which are reused constantly.
This other PR that also does ADADELTA is an example of what I mention: https://github.com/torch/optim/pull/55/files

**Comment by urikz** [comment]
@soumith , LoL, I completely failed the mission =) Alright, so I fixed the memory issue (at least I think so). Feel free to merge either this one or other PRs - I'm fine with both ways.

File: adadelta.lua [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    local decay = config.decayRate or 0.9
+    local eps = config.offset or 1e-6
+
+    assert(decay >= 0 and decay <= 1, "Wrong value for decay rate: " .. decay)
+    assert(eps >= 0, "Wrong value for offset constant: " .. eps)
+
+    local fx, dfdx = opfunc(x)
+
+    if not state.accSquaredDeltaWithDecay then
+        state.accSquaredDeltaWithDecay =
```
Comment (koraykv) [inline adadelta.lua]: one line please

File: adadelta.lua [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            torch.Tensor():typeAs(x):resizeAs(x):zero()
+        state.accSquaredGradsWithDecay =
```
Comment (koraykv) [inline adadelta.lua]: also single line

File: adadelta.lua [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            torch.Tensor():typeAs(dfdx):resizeAs(dfdx):zero()
+    end
+
+    state.accSquaredGradsWithDecay:mul(decay):addcmul(1.0 - decay, dfdx, dfdx)
+
+    if not state.prev_rms_delta then
```
Comment (koraykv) [inline adadelta.lua]: you do not need another if statement, you can just move these into the previous if statement

File: adadelta.lua [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        state.prev_rms_delta = torch.Tensor()
+            :typeAs(state.accSquaredDeltaWithDecay)
+            :resizeAs(state.accSquaredDeltaWithDecay)
+        state.rms_grad = torch.Tensor()
+            :typeAs(state.accSquaredGradsWithDecay)
+            :resizeAs(state.accSquaredGradsWithDecay)
+    end
+
+    state.prev_rms_delta:copy(state.accSquaredDeltaWithDecay):add(eps):sqrt()
```
Comment (koraykv) [inline adadelta.lua]: rename state.prev_rms_delta to state.rmsDelta
rename state.rms_grad to state.rmsGrad

**Comment by koraykv** [comment]
By the way, there is another pull request from @ketranm (https://github.com/torch/optim/pull/55). Maybe it is best to decide which one to go with, but at this moment, both needs some minor changes.

**Comment by urikz** [comment]
@koraykv, Thanks for the comments - I updated the code.

**Comment by koraykv** [comment]
I am going to close this request, since I have already merged the other one. Thanks for the implementation. Please keep them coming!

**Comment by urikz** [comment]
@koraykv , sure! =)

**Pull request closed without merging**